carla/data/causal_model/synthethic_data.py

```diff
@@ -0,0 +1,149 @@
+from typing import List
+
+import numpy as np
+import pandas as pd
+
+from carla.data.api import Data
+
+
+def _get_noise_string(node):
+    if not node[0] == "x":
+        raise ValueError
+    return "u" + node[1:]
+
+
+def _create_synthetic_data(scm, num_samples):
+    """
+    Generate synthetic data.
+
+    Parameters
+    ----------
+    scm : CausalModel
+        Structural causal model
+    num_samples : int
+        Number of samples in the dataset
+
+    """
+
+    exogenous_variables = np.concatenate(
+        [
+            np.array(scm.noise_distributions[node].sample(num_samples)).reshape(-1, 1)
+            for node in scm.get_topological_ordering("exogenous")
+        ],
+        axis=1,
+    )
+    exogenous_variables = pd.DataFrame(
+        exogenous_variables, columns=scm.get_topological_ordering("exogenous")
+    )
+
+    endogenous_variables = exogenous_variables.copy()
+    endogenous_variables = endogenous_variables.rename(
+        columns=dict(
+            zip(
+                scm.get_topological_ordering("exogenous"),
+                scm.get_topological_ordering("endogenous"),
+            )
+        )
+    )
+    # used later to make sure parents are populated when computing children
+    endogenous_variables.loc[:] = np.nan
+    for node in scm.get_topological_ordering("endogenous"):
+        parents = scm.get_parents(node)
+        if endogenous_variables.loc[:, list(parents)].isnull().values.any():
+            raise ValueError(
+                "parents in endogenous_variables should already be occupied"
+            )
+        endogenous_variables[node] = scm.structural_equations_np[node](
+            exogenous_variables[_get_noise_string(node)],
+            *[endogenous_variables[p] for p in parents],
+        )
+
+    # fix a hyperplane
```

Comment:
I see why a hyperplane would need to be fixed, but I do think that there are more intuitive ways than how it's currently written. We can keep this as is rn, but should consider simplifying affairs in the future.